scheduler/src/cook/log_structured.clj

```diff
@@ -0,0 +1,52 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+
+(ns cook.log-structured
+  (:require
+    [clojure.data.json :as json]
+    [clojure.tools.logging :as log]))
+
+(defmacro logs
+  "Logs a structured message at the specified level.
+  Supports specifying a custom logger namespace as an optional parameter."
+  ([level data metadata]
+   `log/log ~level (json/write-str (assoc ~metadata :data ~data)))
+  ([level data metadata logger-ns]
```

Comment:
Do we need this anymore? (My thoughts are it sets up a dangerous pattern. Having a trustable namespace is very important, IMO. It lets you trace down where things are comign from. Things won't e.g., go haywire with code moved between namespaces.

Reply:
Also, I think we can supersede this with standardization of what we always put in the metadata dictionary. E.g., we always have the relevant keys, where applicable:
  :compute-cluster-name
  :instance-id
  :job-id
  :module  ; Abstracts over keys in multiple namespaces, e.g., REST, or K8S.
  :user

Reply:
Actually, really big toplevel comment: I think there needs to be a toplevel comment along the lines of this in here, to indicate he desired usecases and how this is intended to be used:


> 
> ;; This file implements structured logging. A structured log line consists of a human readable string "Committing Blah to Blah", followed by a dictionary. 
> (structured-log/info "Launching task " {:instance_id ... ... ...}"
> 
> ;; Within the dictionary, we are standardizing the following key names to be used, where relevant:
> ; :compute-cluster-name
> ; :user 
> ; :instance-id
> ; :job-id
> 

Can you add something along these lines and then if we all think its looks reasonable, we can move along. If not, a quick discussion to figure it out?

Reply:
Added top-level comment.